luomu-common/src/tagged_macaddr.rs

```diff
@@ -0,0 +1,213 @@
+use crate::{MacAddr, TagError};
+
+// Size of our tag stack. 64bit integer can store up to five VLAN tags.
+type TagStack = u64;
+
+/// A [MacAddr] with additional support for storing stack of VLAN IDs.
+///
+/// Tags are stored as a stack where the outermost tag should be pushed first
+/// and popped last (aka LIFO). There's enough room to store up to five VLAN
+/// IDs.
+///
+/// ```rust
+/// use luomu_common::{MacAddr, TaggedMacAddr};
+///
+/// let mut mac = TaggedMacAddr::new("11:22:33:aa:bb:cc".parse().unwrap());
+/// assert_eq!(mac.pop_tag(), None);
+///
+/// mac.push_tag(42);
+/// mac.push_tag(1337);
+/// assert_eq!(mac.pop_tag(), Some(1337));
+/// assert_eq!(mac.pop_tag(), Some(42));
+/// assert_eq!(mac.pop_tag(), None);
+/// ```
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TaggedMacAddr {
+    mac: MacAddr,
+    tag_stack: TagStack,
+}
+
+impl TaggedMacAddr {
+    /// Construct new [TaggedMacAddr].
+    pub const fn new(mac: MacAddr) -> Self {
```

Comment:
Should we ensure there is no vlan tag already on the `MacAddr` and if there is, maybe push it onto the stack.

Reply:
We have two places to store tags: One can be in `MacAddr` and five more in `TaggedMacAddr`. This can be used for example by storing the customer tag in `MacAddr` and service tags in `TaggedMacAddr` so we can store up to six tags and one MAC address in 16 bytes.